File: src/main/java/no/novari/flyt/archive/gateway/dispatch/mapping/JournalpostMappingService.java
package no.novari.flyt.archive.gateway.dispatch.mapping;

import no.fint.model.resource.Link;
import no.fint.model.resource.arkiv.noark.JournalpostResource;
import no.novari.flyt.archive.gateway.dispatch.model.instance.JournalpostDto;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.UUID;

@Service
public class JournalpostMappingService {

    private final SkjermingMappingService skjermingMappingService;
    private final KorrespondansepartMappingService korrespondanseparMappingService;
    private final DokumentbeskrivelseMappingService dokumentbeskrivelseMappingService;

    public JournalpostMappingService(
            SkjermingMappingService skjermingMappingService,
            KorrespondansepartMappingService korrespondanseparMappingService,
            DokumentbeskrivelseMappingService dokumentbeskrivelseMappingService
    ) {
        this.skjermingMappingService = skjermingMappingService;
        this.korrespondanseparMappingService = korrespondanseparMappingService;
        this.dokumentbeskrivelseMappingService = dokumentbeskrivelseMappingService;
    }

    public JournalpostResource toJournalpostResource(
            JournalpostDto journalpostDto,
            Map<UUID, Link> fileArchiveLinkPerFileId
    ) {
        JournalpostResource journalpostResource = new JournalpostResource();

        journalpostDto.getTittel().ifPresent(journalpostResource::setTittel);
        journalpostDto.getOffentligTittel().ifPresent(journalpostResource::setOffentligTittel);
        journalpostDto.getJournalstatus().map(Link::with).ifPresent(journalpostResource::addJournalstatus);
        journalpostDto.getTilgangsgruppe().map(Link::with).ifPresent(journalpostResource::addTilgangsgruppe);
        journalpostDto.getSaksbehandler().map(Link::with).ifPresent(journalpostResource::addSaksbehandler);
        journalpostDto.getJournalposttype().map(Link::with).ifPresent(journalpostResource::addJournalposttype);
        journalpostDto.getAdministrativEnhet().map(Link::with).ifPresent(journalpostResource::addAdministrativEnhet);

        journalpostDto.getSkjerming()
                .map(skjermingMappingService::toSkjermingResource)
                .ifPresent(journalpostResource::setSkjerming);

        journalpostDto.getKorrespondansepart()
                .map(korrespondanseparMappingService::toKorrespondansepartResource)
                .ifPresent(journalpostResource::setKorrespondansepart);

        journalpostDto.getDokumentbeskrivelse()
                .map(dokumentBeskrivelseDtos -> dokumentbeskrivelseMappingService.toDokumentbeskrivelseResource(
                        dokumentBeskrivelseDtos, fileArchiveLinkPerFileId
                ))
                .ifPresent(journalpostResource::setDokumentbeskrivelse);

        return journalpostResource;
    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/mapping/JournalpostMappingService.kt
package no.novari.flyt.archive.gateway.dispatch.mapping

import no.novari.fint.model.resource.Link
import no.novari.fint.model.resource.arkiv.noark.JournalpostResource
import no.novari.flyt.archive.gateway.dispatch.model.instance.JournalpostDto
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class JournalpostMappingService(
    private val skjermingMappingService: SkjermingMappingService,
    private val korrespondansepartMappingService: KorrespondansepartMappingService,
    private val dokumentbeskrivelseMappingService: DokumentbeskrivelseMappingService,
) {
    fun toJournalpostResource(
        journalpostDto: JournalpostDto,
        fileArchiveLinkPerFileId: Map<UUID, Link>,
    ): JournalpostResource =
        JournalpostResource().apply {
            journalpostDto.tittel?.let(::setTittel)
            journalpostDto.offentligTittel?.let(::setOffentligTittel)
            journalpostDto.journalstatus?.let(Link::with)?.let(::addJournalstatus)
            journalpostDto.tilgangsgruppe?.let(Link::with)?.let(::addTilgangsgruppe)
            journalpostDto.saksbehandler?.let(Link::with)?.let(::addSaksbehandler)
            journalpostDto.journalposttype?.let(Link::with)?.let(::addJournalposttype)
            journalpostDto.administrativEnhet?.let(Link::with)?.let(::addAdministrativEnhet)
            journalpostDto.skjerming?.let(skjermingMappingService::toSkjermingResource)?.let(::setSkjerming)
            journalpostDto.korrespondansepart
                ?.let(korrespondansepartMappingService::toKorrespondansepartResource)
                ?.let(::setKorrespondansepart)
            journalpostDto.dokumentbeskrivelse
                ?.let { dokumentbeskrivelseMappingService.toDokumentbeskrivelseResource(it, fileArchiveLinkPerFileId) }
                ?.let(::setDokumentbeskrivelse)
        }
}
